### ISSUE TITLE
Sharing word embedding weights between source and target

### ISSUE THREAD

**Issue by atebbifakhr** [OpenNMT/OpenNMT-tf#143]
Hi,

I'm using the multi-source model, since there are three different vocabularies, the number of parameters are huge. It would be nice if I could share embedding weights between sources or source and target. Do you have any plan for adding this feature? I can contribute if you tell me which files should be changed.

**Comment by guillaumekln** [comment]
Hello,

I would like to add this feature (plus the ability to share target embeddings and the softmax weights) but can't give any ETA for that at the moment.

The word embedding variable is created at the beginning of [this function](https://github.com/OpenNMT/OpenNMT-tf/blob/v1.4.1/opennmt/inputters/text_inputter.py#L364). A basic implementation would simply pass the top-level scope and create the variable under that scope using `reuse`.

**Comment by wanghm92** [comment]
Hi! I just ran into this issue and thanks for raising it and the hint! 
May I ask if this is going to be implemented any time soon? 
Or could you please give more hint on where should I add another tf.variable_scope() to make the embeddings reusable? Thanks!

**Comment by guillaumekln** [comment]
I started supporting embeddings sharing in the PR above. For now, it only supports a simple source/target word embeddings sharing (the most standard use case).

Let's keep this issue open to keep track of the support of additional sharing levels:

* [ ] share between source embeddings (as requested in the opening message)
* [ ] share target word embeddings and softmax weights
* [ ] share all embeddings

**Comment by bidai541** [comment]
@guillaumekln  Is there an entry point for this parameter of sharing embedding?  I found that the init function of catalog.transformer do not include "share_embeddings"  while this parameter in superclass opennmt.models.transformer is fixed to 0.

**Comment by guillaumekln** [comment]
You should define a new model configuration to enable this parameter (see existing configurations in `config/models/`).

For example create a file `transformer_shared_embedding.py`:

```python
import tensorflow as tf
import opennmt as onmt

def model():
  return onmt.models.Transformer(
      source_inputter=onmt.inputters.WordEmbedder(
          vocabulary_file_key="source_words_vocabulary",
          embedding_size=512),
      target_inputter=onmt.inputters.WordEmbedder(
          vocabulary_file_key="target_words_vocabulary",
          embedding_size=512),
      num_layers=6,
      num_units=512,
      num_heads=8,
      ffn_inner_dim=2048,
      dropout=0.1,
      attention_dropout=0.1,
      relu_dropout=0.1,
      share_embeddings=onmt.models.EmbeddingsSharingLevel.SOURCE_TARGET_INPUT)
```

and reference it when starting the training:

```bash
onmt-main train_and_eval --model transformer_shared_embedding.py [...]
```

**Comment by bidai541** [comment]
Hi @guillaumekln 
I have got another problem.
1. while train_and_eval , the program will quit after the first eval step 
2. while run onmt-main eval, the loss is right but the BLEU is 0.0000
Looking forward to replies!

```
train:
  batch_size: 4096
  batch_type: tokens
  bucket_width: 1
  save_checkpoints_steps: 1000
  keep_checkpoint_max: 5
  save_summary_steps: 100
  train_steps: 300000
  maximum_features_length: 30
  maximum_labels_length: 30

  # Uniformly shuffle accross the training data.
  # Set a fixed value if you are running low on system memory.
  sample_buffer_size: -1

eval:
  batch_size: 32
  eval_delay: 3600  # Every 10 hour
  external_evaluators: BLEU
```

**Comment by guillaumekln** [comment]
Hi,

1. What is the error when using `train_and_eval`? Is it related to this feature?
2. What is the reported loss value? It can happen that the BLEU score is 0 in the early training stages, or when there is an issue with the data.

**Comment by bidai541** [comment]
The first problem may be due to memory increasing while changing train and evel steps (no any log and error, just stoped......  I loaded all the dataset at once which is more than a hundred million). It's ok while setting buffer_size.   
The second problem is that:  one model saved at step 10001, loss=0.1 bleu=3.0 at train_and_eval, while running eval alone at the same dataset and saved model, loss=0.1 , blue=0.0.  Maybe something wrong with myself,  I will check again.  
Thanks

**Comment by guillaumekln** [comment]
This issue can finally be closed thanks to the PR above. All sharing combinations are now implemented.

**Comment by atebbifakhr** [comment]
It is a useful feature, thanks a lot!
